File: chat/lib/llm/azure-openai-client.ts
/**
 * Pure Azure OpenAI LLM Client
 * Handles only LLM interactions without MCP coupling
 */

import { AzureOpenAI } from 'openai';
import { DefaultAzureCredential, getBearerTokenProvider } from '@azure/identity';
import { AzureOpenAIConfig } from '../mcp/types';
import { z } from 'zod';

// Zod schemas for validation
const llmMessageSchema = z.object({
  role: z.enum(['user', 'assistant', 'system']),
  content: z.string()
});

const llmToolSchema = z.object({
  name: z.string(),
  description: z.string(),
  parameters: z.object({
    type: z.literal('object'),
    properties: z.record(z.any()),
    required: z.array(z.string()).optional()
  })
});

const llmToolCallSchema = z.object({
  id: z.string(),
  name: z.string(),
  arguments: z.record(z.any())
});

const llmChatOptionsSchema = z.object({
  messages: z.array(llmMessageSchema).min(1),
  tools: z.array(llmToolSchema).optional(),
  maxTokens: z.number().positive().optional(),
  temperature: z.number().min(0).max(2).optional(),
  stream: z.boolean().optional()
});

export interface LLMMessage {
  role: 'user' | 'assistant' | 'system';
  content: string;
}

export interface LLMTool {
  name: string;
  description: string;
  parameters: {
    type: 'object';
    properties: Record<string, any>;
    required?: string[];
  };
}

export interface LLMToolCall {
  id: string;
  name: string;
  arguments: Record<string, any>;
}

export interface LLMChatOptions {
  messages: LLMMessage[];
  tools?: LLMTool[];
  maxTokens?: number;
  temperature?: number;
  stream?: boolean;
}

export interface LLMChatResponse {
  content?: string;
  toolCalls?: LLMToolCall[];
  finishReason?: 'stop' | 'tool_calls' | 'length' | 'content_filter';
  usage?: {
    promptTokens: number;
    completionTokens: number;
    totalTokens: number;
  };
}

class LLMValidationError extends Error {
  constructor(message: string, public details?: any) {
    super(message)
    this.name = 'LLMValidationError'
  }
}

export class AzureOpenAIClient {
  private client: AzureOpenAI | null = null;
  private config: AzureOpenAIConfig;

  constructor(config: AzureOpenAIConfig) {
    this.config = config;
  }

  /**
   * Initialize Azure OpenAI client with appropriate authentication
   */
  async initialize(): Promise<void> {
    try {
      if (this.config.useAzureAD) {
        // Use Azure AD authentication (managed identity or service principal)
        const credential = new DefaultAzureCredential();
        const scope = "https://cognitiveservices.azure.com/.default";
        const azureADTokenProvider = getBearerTokenProvider(credential, scope);
        
        this.client = new AzureOpenAI({
          azureADTokenProvider,
          deployment: this.config.deploymentName,
          apiVersion: this.config.apiVersion || '2025-01-01-preview'
        });
      } else if (this.config.apiKey) {
        // Use API key authentication
        this.client = new AzureOpenAI({
          apiKey: this.config.apiKey,
          endpoint: this.config.endpoint,
          deployment: this.config.deploymentName,
          apiVersion: this.config.apiVersion || '2025-01-01-preview'
        });
      } else {
        throw new Error('Either Azure AD authentication or API key must be configured');
      }

      console.log('✅ Azure OpenAI LLM client initialized successfully');
    } catch (error) {
      console.error('❌ Failed to initialize Azure OpenAI LLM client:', error);
      throw new Error(`Azure OpenAI LLM client initialization failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Create chat completion with optional tool calling
   */
  async createChatCompletion(options: LLMChatOptions): Promise<LLMChatResponse> {
    if (!this.client) {
      throw new LLMValidationError('Azure OpenAI client not initialized');
    }

    try {
      // Validate options with Zod
      const validatedOptions = llmChatOptionsSchema.parse(options);
      
      const messages = this.formatMessages(validatedOptions.messages);
      const tools = validatedOptions.tools ? this.formatTools(validatedOptions.tools) : undefined;

      const response = await this.client.chat.completions.create({
        model: this.config.deploymentName,
        messages,
        tools,
        tool_choice: tools ? 'auto' : undefined,
        max_tokens: validatedOptions.maxTokens || 2000,
        temperature: validatedOptions.temperature || 0.7,
        stream: false
      });

      const choice = response.choices[0];
      if (!choice) {
        throw new Error('No response choice received from Azure OpenAI');
      }

      return {
        content: choice.message?.content || undefined,
        toolCalls: this.parseToolCalls(choice.message?.tool_calls || []),
        finishReason: this.mapFinishReason(choice.finish_reason),
        usage: response.usage ? {
          promptTokens: response.usage.prompt_tokens,
          completionTokens: response.usage.completion_tokens,
          totalTokens: response.usage.total_tokens
        } : undefined
      };
    } catch (error) {
      console.error('❌ Azure OpenAI API error:', error);
      throw new Error(`Chat completion failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Create streaming chat completion
   */
  async *createStreamingChatCompletion(options: LLMChatOptions): AsyncGenerator<LLMChatResponse, void, unknown> {
    if (!this.client) {
      throw new LLMValidationError('Azure OpenAI client not initialized');
    }

    try {
      // Validate options with Zod
      const validatedOptions = llmChatOptionsSchema.parse(options);
      
      const messages = this.formatMessages(validatedOptions.messages);
      const tools = validatedOptions.tools ? this.formatTools(validatedOptions.tools) : undefined;

      const stream = await this.client.chat.completions.create({
        model: this.config.deploymentName,
        messages,
        tools,
        tool_choice: tools ? 'auto' : undefined,
        max_tokens: validatedOptions.maxTokens || 2000,
        temperature: validatedOptions.temperature || 0.7,
        stream: true
      });

      // Accumulator for tool calls across chunks
      const accumulatedToolCalls: { [index: number]: { id: string; function: { name: string; arguments: string } } } = {};

      for await (const chunk of stream) {
        const choice = chunk.choices[0];
        if (!choice) continue;

        const delta = choice.delta;
        if (!delta) continue;

        // Handle content chunks
        if (delta.content) {
          yield {
            content: delta.content
          } as LLMChatResponse;
        }

        // Handle tool calls - accumulate them as they stream in
        if (delta.tool_calls && delta.tool_calls.length > 0) {
          for (const toolCall of delta.tool_calls) {
            if (toolCall.index !== undefined) {
              const index = toolCall.index;
              
              // Initialize tool call accumulator if needed
              if (!accumulatedToolCalls[index]) {
                accumulatedToolCalls[index] = {
                  id: '',
                  function: { name: '', arguments: '' }
                };
              }
              
              // Accumulate the tool call data
              if (toolCall.id) {
                accumulatedToolCalls[index].id = toolCall.id;
              }
              if (toolCall.function?.name) {
                accumulatedToolCalls[index].function.name = toolCall.function.name;
              }
              if (toolCall.function?.arguments) {
                accumulatedToolCalls[index].function.arguments += toolCall.function.arguments;
              }
            }
          }
        }

        // Handle finish reason
        if (choice.finish_reason) {
          // If we have accumulated tool calls, yield them before finish reason
          if (Object.keys(accumulatedToolCalls).length > 0) {
            const toolCallsArray = Object.values(accumulatedToolCalls);
            const parsedToolCalls = this.parseToolCalls(toolCallsArray);
            if (parsedToolCalls.length > 0) {
              yield {
                toolCalls: parsedToolCalls
              } as LLMChatResponse;
            }
          }
          
          yield {
            finishReason: this.mapFinishReason(choice.finish_reason)
          } as LLMChatResponse;
          break;
        }
      }
    } catch (error) {
      console.error('❌ Azure OpenAI streaming error:', error);
      throw new Error(`Streaming chat completion failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Check if client is ready
   */
  isReady(): boolean {
    return this.client !== null;
  }

  /**
   * Format messages for Azure OpenAI API
   */
  private formatMessages(messages: LLMMessage[]): any[] {
    return messages.map(msg => ({
      role: msg.role,
      content: msg.content
    }));
  }

  /**
   * Format tools for Azure OpenAI API
   */
  private formatTools(tools: LLMTool[]): any[] {
    return tools.map(tool => ({
      type: 'function',
      function: {
        name: tool.name,
        description: tool.description,
        parameters: tool.parameters
      }
    }));
  }

  /**
   * Parse tool calls from Azure OpenAI response with validation
   */
  private parseToolCalls(toolCalls: any[]): LLMToolCall[] {
    return toolCalls.map(tc => {
      try {
        // Parse arguments - with 2025-01-01-preview, arguments are more reliable
        const parsedArguments = tc.function?.arguments 
          ? (typeof tc.function.arguments === 'string' 
             ? JSON.parse(tc.function.arguments) 
             : tc.function.arguments)
          : {};
        
        const toolCall = {
          id: tc.id || '',
          name: tc.function?.name || '',
          arguments: parsedArguments
        };
        
        // Validate with Zod
        return llmToolCallSchema.parse(toolCall);
      } catch (error) {
        console.error('❌ Invalid tool call data:', error, tc);
        // Return a safe fallback
        return {
          id: tc.id || 'unknown',
          name: tc.function?.name || 'unknown',
          arguments: {}
        };
      }
    }).filter(tc => tc.name && tc.name !== 'unknown'); // Filter out invalid tool calls
  }

  /**
   * Map Azure OpenAI finish reason to our format
   */
  private mapFinishReason(reason: string | null): 'stop' | 'tool_calls' | 'length' | 'content_filter' {
    switch (reason) {
      case 'stop': return 'stop';
      case 'tool_calls': return 'tool_calls';
      case 'length': return 'length';
      case 'content_filter': return 'content_filter';
      default: return 'stop';
    }
  }
}

File: agent/azure_openai_client.py
#!/usr/bin/env python3
"""
Azure OpenAI Client for Agent System

Pure Azure OpenAI LLM Client that handles only LLM interactions,
supporting both Azure AD and API key authentication.
"""

import asyncio
import logging
from typing import List, Dict, Any, Optional, Union
from dataclasses import dataclass

from openai import AsyncAzureOpenAI
from azure.identity.aio import DefaultAzureCredential
from azure.core.credentials import AccessToken

from azure_openai_config import AzureOpenAIConfig, load_azure_openai_config

logger = logging.getLogger(__name__)

@dataclass
class LLMMessage:
    """LLM message structure"""
    role: str  # 'user', 'assistant', 'system'
    content: str

@dataclass
class LLMChatOptions:
    """LLM chat completion options"""
    messages: List[LLMMessage]
    max_tokens: Optional[int] = 2000
    temperature: Optional[float] = 0.7
    stream: bool = False

@dataclass
class LLMChatResponse:
    """LLM chat completion response"""
    content: Optional[str] = None
    finish_reason: Optional[str] = None
    usage: Optional[Dict[str, int]] = None

class LLMValidationError(Exception):
    """Custom exception for LLM validation errors"""
    def __init__(self, message: str, details: Any = None):
        super().__init__(message)
        self.details = details

class AzureOpenAIClient:
    """Azure OpenAI client with support for both authentication methods"""
    
    def __init__(self, config: Optional[AzureOpenAIConfig] = None):
        self.config = config or load_azure_openai_config()
        self.client: Optional[AsyncAzureOpenAI] = None
        self._azure_credential: Optional[DefaultAzureCredential] = None
        
        if not self.config:
            raise LLMValidationError("Azure OpenAI configuration not found")
    
    async def initialize(self) -> None:
        """Initialize Azure OpenAI client with appropriate authentication"""
        try:
            if self.config.use_azure_ad:
                # Use Azure AD authentication (managed identity or service principal)
                self._azure_credential = DefaultAzureCredential()
                
                # For Azure AD, we need to get tokens manually and create client without explicit auth
                self.client = AsyncAzureOpenAI(
                    azure_endpoint=self.config.endpoint,
                    api_version=self.config.api_version,
                    azure_ad_token_provider=self._get_azure_ad_token
                )
            else:
                # Use API key authentication
                if not self.config.api_key:
                    raise LLMValidationError("API key is required when not using Azure AD authentication")
                
                self.client = AsyncAzureOpenAI(
                    api_key=self.config.api_key,
                    azure_endpoint=self.config.endpoint,
                    api_version=self.config.api_version
                )
            
            logger.info("✅ Azure OpenAI LLM client initialized successfully")
            
        except Exception as error:
            logger.error(f"❌ Failed to initialize Azure OpenAI LLM client: {error}")
            raise LLMValidationError(f"Azure OpenAI LLM client initialization failed: {error}")
    
    async def _get_azure_ad_token(self) -> str:
        """Get Azure AD token for authentication"""
        if not self._azure_credential:
            raise LLMValidationError("Azure credential not initialized")
        
        scope = "https://cognitiveservices.azure.com/.default"
        token = await self._azure_credential.get_token(scope)
        return token.token
    
    async def create_chat_completion(self, options: LLMChatOptions) -> LLMChatResponse:
        """Create chat completion with the configured model"""
        if not self.client:
            raise LLMValidationError("Azure OpenAI client not initialized")
        
        try:
            # Format messages for OpenAI API
            messages = [
                {"role": msg.role, "content": msg.content}
                for msg in options.messages
            ]
            
            response = await self.client.chat.completions.create(
                model=self.config.deployment_name,
                messages=messages,
                max_tokens=options.max_tokens,
                temperature=options.temperature,
                stream=options.stream
            )
            
            choice = response.choices[0] if response.choices else None
            if not choice:
                raise LLMValidationError("No response choice received from Azure OpenAI")
            
            return LLMChatResponse(
                content=choice.message.content,
                finish_reason=choice.finish_reason,
                usage={
                    "prompt_tokens": response.usage.prompt_tokens if response.usage else 0,
                    "completion_tokens": response.usage.completion_tokens if response.usage else 0,
                    "total_tokens": response.usage.total_tokens if response.usage else 0
                } if response.usage else None
            )
            
        except Exception as error:
            logger.error(f"❌ Azure OpenAI API error: {error}")
            raise LLMValidationError(f"Chat completion failed: {error}")
    
    async def create_streaming_chat_completion(self, options: LLMChatOptions):
        """Create streaming chat completion"""
        if not self.client:
            raise LLMValidationError("Azure OpenAI client not initialized")
        
        try:
            # Format messages for OpenAI API
            messages = [
                {"role": msg.role, "content": msg.content}
                for msg in options.messages
            ]
            
            stream = await self.client.chat.completions.create(
                model=self.config.deployment_name,
                messages=messages,
                max_tokens=options.max_tokens,
                temperature=options.temperature,
                stream=True
            )
            
            async for chunk in stream:
                if chunk.choices and chunk.choices[0].delta.content:
                    yield LLMChatResponse(
                        content=chunk.choices[0].delta.content,
                        finish_reason=chunk.choices[0].finish_reason
                    )
                    
        except Exception as error:
            logger.error(f"❌ Azure OpenAI streaming API error: {error}")
            raise LLMValidationError(f"Streaming chat completion failed: {error}")
    
    def is_ready(self) -> bool:
        """Check if client is ready for use"""
        return self.client is not None
    
    async def close(self) -> None:
        """Clean up resources"""
        if self.client:
            await self.client.close()
        if self._azure_credential:
            await self._azure_credential.close()

# Global client instance
_global_client: Optional[AzureOpenAIClient] = None

async def get_azure_openai_client() -> AzureOpenAIClient:
    """Get or create the global Azure OpenAI client instance"""
    global _global_client
    
    if _global_client is None:
        _global_client = AzureOpenAIClient()
        await _global_client.initialize()
    
    return _global_client

async def call_llm(
    system_prompt: str, 
    user_prompt: str, 
    max_tokens: int = 1000,
    temperature: float = 0.3
) -> str:
    """
    Convenience function for simple LLM calls
    Compatible with the existing base_agent.call_llm interface
    """
    client = await get_azure_openai_client()
    
    options = LLMChatOptions(
        messages=[
            LLMMessage(role="system", content=system_prompt),
            LLMMessage(role="user", content=user_prompt)
        ],
        max_tokens=max_tokens,
        temperature=temperature
    )
    
    response = await client.create_chat_completion(options)
    return response.content or "" 